service/test_handlers.py

```diff
@@ -1,10 +1,257 @@
+    Postgresql.clear_cache()
+
+
+class M2HandlersIntegrationCase(unittest.TestCase):
+
+    def setUp(self):
+        self.pg = Postgresql()
+        os.environ['pg-host'] = self.pg.dsn()['host']
+        os.environ['pg-port'] = str(self.pg.dsn()['port'])
+        os.environ['pg-database'] = self.pg.dsn()['database']
+        os.environ['pg-user'] = self.pg.dsn()['user']
+
+    def tearDown(self):
+        self.pg.stop()
+
+    def test_handle_build(self):
+        """ handle branch should save the service, branch and iteration """
+        # run SUT
+        handle_build(
+            'mock_service',
+            'mock_branch',
+            'mock_commitabc112',
+            'us.gcr.io/mock-image:v0.1',
+        )
+
+        # confirm
+        cursor = get_cursor()
+        cursor.execute(
+            "select service_name, branch_name, commit_hash, image_name\n" + \
+            "  from service\n" + \
+            "  join branch using (service_id)\n" + \
+            "  join iteration using (branch_id)\n" + \
+            " where image_name='us.gcr.io/mock-image:v0.1'"
+        )
+        results = cursor.fetchall()
+        self.assertEqual(len(results), 1)
+        self.assertEqual(
+            results[0],
+            ('mock_service',
+             'mock_branch',
+             'mock_commitabc112',
+             'us.gcr.io/mock-image:v0.1')
+        )
+        cursor.close()
+
+    @given(
+        commit_hash=text(max_size=99),
+        image_name=text(max_size=99)
+    )
+    @example(
+        commit_hash='anything',
+        image_name='something\x00weird',
+    )
+    @example(
+        commit_hash='\x0b\n\x05\U000beb17\x07',
+        image_name='\x00',
+    )
+    def test_save_idempotent_by_uniqueness(
+            self,
+            commit_hash,
+            image_name,
+    ):
+        """
+        iterations should be idempotent, even though no all of their
+        columns are included in uniqueness constraints
+
+        anything with the null character will be truncated at the null character
+        because this is what postgresql does.
+
+        """
+        print(commit_hash)
+        print(image_name)
+
+        cursor = get_cursor()
+        service_id = save(
+            cursor,
+            'service',          # table name
+            ['service_name'],   # unique columns
+            ['service_name'],   # all columns
+            ['mock_service'],   # values to insert
+        )
+        branch_id = save(
+            cursor,
+            'branch',                        # table name
+            ['service_id', 'branch_name'],   # unique columns
+            ['service_id', 'branch_name'],   # all columns
+            [ service_id , 'mock_branch'],   # values to insert
+        )
+
+        """
+        If someone tries to save a build name to a branch-commit pair that
+        exists the existing build name will be used and the new build name
+        being saved will be ignored.
+        """
```

Comment:
@jessebmiller Add a note here about how the property testing might give us an commit that we've already seen paired with a different image than previously provided with that commit, necessitating lines 137 - 150